discovery/store.go

```diff
@@ -0,0 +1,379 @@
+		PresentationRaw:        presentation.Raw(),
+		PresentationExpiration: presentation.JWT().Expiration().Unix(),
+	}
+
+	for _, currCred := range presentation.VerifiableCredential {
+		var credentialType *string
+		for _, currType := range currCred.Type {
+			if currType.String() != "VerifiableCredential" {
+				credentialType = new(string)
+				*credentialType = currType.String()
+				break
+			}
+		}
+		if len(currCred.CredentialSubject) != 1 {
+			return nil, errors.New("credential must contain exactly one subject")
+		}
+
+		newCredential := credentialRecord{
+			ID:                  uuid.NewString(),
+			PresentationID:      newPresentation.ID,
+			CredentialID:        currCred.ID.String(),
+			CredentialIssuer:    currCred.Issuer.String(),
+			CredentialSubjectID: credentialSubjectID.String(),
+			CredentialType:      credentialType,
+		}
+		// Store credential's properties
+		keys, values := indexJSONObject(currCred.CredentialSubject[0].(map[string]interface{}), nil, nil, "credentialSubject")
+		for i, key := range keys {
+			if key == "credentialSubject.id" {
+				// present as column, don't index
+				continue
+			}
+			newCredential.Properties = append(newCredential.Properties, credentialPropertyRecord{
+				CredentialID: newCredential.ID,
+				Key:          key,
+				Value:        values[i],
+			})
+		}
+		newPresentation.Credentials = append(newPresentation.Credentials, newCredential)
+	}
+	return &newPresentation, nil
+}
+
+func (s *sqlStore) get(serviceID string, startAt Timestamp) ([]vc.VerifiablePresentation, *Timestamp, error) {
+	var rows []presentationRecord
+	err := s.db.Order("lamport_timestamp ASC").Find(&rows, "service_id = ? AND lamport_timestamp > ?", serviceID, int(startAt)).Error
+	if err != nil {
+		return nil, nil, fmt.Errorf("query service '%s': %w", serviceID, err)
+	}
+	timestamp := startAt
+	presentations := make([]vc.VerifiablePresentation, 0, len(rows))
+	for _, row := range rows {
+		presentation, err := vc.ParseVerifiablePresentation(row.PresentationRaw)
+		if err != nil {
+			return nil, nil, fmt.Errorf("parse presentation '%s' of service '%s': %w", row.PresentationID, serviceID, err)
+		}
+		presentations = append(presentations, *presentation)
+		timestamp = Timestamp(row.LamportTimestamp)
+	}
+	return presentations, &timestamp, nil
+}
+
+func (s *sqlStore) search(serviceID string, query map[string]string) ([]vc.VerifiablePresentation, error) {
+	propertyColumns := map[string]string{
+		"id":                   "cred.credential_id",
+		"issuer":               "cred.credential_issuer",
+		"type":                 "cred.credential_type",
+		"credentialSubject.id": "cred.credential_subject_id",
+	}
+
+	stmt := s.db.Model(&presentationRecord{}).
+		Where("service_id = ?", serviceID).
+		Joins("inner join discovery_credential cred ON cred.presentation_id = discovery_presentation.id")
+	numProps := 0
+	for jsonPath, value := range query {
+		if value == "*" {
+			continue
+		}
+		// sort out wildcard mode
```

Comment:
todo?

Reply:
no, short explanation of the code below. Will elaborate a bit on it.